### PR TITLE
[Vue] migrate scheduled reports management from twig/angularjs to vue

### PR DESCRIPTION
**Pull request by diosmosis**
### Description:

Changes:
* Migrated scheduled reports management from twig/angularjs to vue.
* Migrated MobileMessaging extension to ScheduledReports to Vue.
* Add redirectOnSuccess option to AjaxHelper.
* Recognize vue-entry-ignore attribute in compileVueEntryComponents() to avoid mounting vue components even if they have vue-entry.
* Add extraProps to compileVueEntryComponents() to pass extra property values to each vue-entry.

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

**Comment by diosmosis** [comment]
Converting to draft for a bit, might be possible to provide full BC.

File: plugins/ScheduledReports/ScheduledReports.php [hunk 1]
```diff
@@ -126,6 +126,37 @@ public function getClientSideTranslationKeys(&$translationKeys)
...
+        $translationKeys[] = 'ScheduledReports_DescriptionOnFirstPage';
+        $translationKeys[] = 'SegmentEditor_ChooseASegment';
+        $translationKeys[] = 'ScheduledReports_WeeklyScheduleHelp';
+        $translationKeys[] = 'ScheduledReports_MonthlyScheduleHelp';
+        $translationKeys[] = 'ScheduledReports_ReportPeriod';
+        $translationKeys[] = 'ScheduledReports_ReportPeriodHelp';
+        $translationKeys[] = 'ScheduledReports_ReportPeriodHelp2';
+        $translationKeys[] = 'ScheduledReports_ReportHour';
+        $translationKeys[] = 'ScheduledReports_ReportType';
+        $translationKeys[] = 'ScheduledReports_ReportFormat';
```
Comment (sgiehl) [inline plugins/ScheduledReports/ScheduledReports.php]: That one seems to be a duplicate
```suggestion
```

File: plugins/ScheduledReports/vue/src/ManageScheduledReport/ManageScheduledReport.vue [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+      AjaxHelper.post(
+        {
+          method: 'ScheduledReports.sendReport',
+        },
+        {
+          idReport,
+          force: true,
+        },
+      ).then(() => {
+        this.fadeInOutSuccessMessage(
```
Comment (sgiehl) [inline plugins/ScheduledReports/vue/src/ManageScheduledReport/ManageScheduledReport.vue:226]: I guess it was already the same before, but `fadeInOutSuccessMessage` currently always does a page reload. At this point that is actually not needed, as no new information could be displayed anyway.

File: plugins/ScheduledReports/vue/src/ManageScheduledReport/ManageScheduledReport.vue [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+          method: isCreate ? 'ScheduledReports.updateReport' : 'ScheduledReports.addReport',
+          period,
+          hour,
+        },
+        apiParameters,
+        {
+          redirectOnSuccess: true,
+        },
+      ).then(() => {
+        this.fadeInOutSuccessMessage(
```
Comment (sgiehl) [inline plugins/ScheduledReports/vue/src/ManageScheduledReport/ManageScheduledReport.vue:378]: That method does a page reload, so guess we could remove the `redirectOnSuccess: true,` above, as the request caused by that gets aborted directly.

File: plugins/ScheduledReports/templates/index.twig [hunk 4]
```diff
@@ -9,19 +9,40 @@
...
+
+{# load the Template.reportParametersScheduledReports event twice, once here outside of vue so any inline scripts
+   will execute. then again in the vue slot so it will be used correctly. This is hack to provide some level of BC. #}
+<div style="display:none" vue-entry-ignore ng-non-bindable>
+    {{ reportParametersScheduledReportsEvent|raw }}
+</div>
+
+<div
+  vue-entry="ScheduledReports.ManageScheduledReport"
+  title="{{ title|default(null)|json_encode|e('html_attr') }}"
```
Comment (sgiehl) [inline plugins/ScheduledReports/templates/index.twig]: Using an attribute named `title` causes a tooltip to be shown everywhere within that div. So might be better to use another name here.

Comment (sgiehl) [reply plugins/ScheduledReports/templates/index.twig]: In addition the `json_encode` here also causes special charts like umlauts to be displayed encoded in the tooltip.
Does Vue always perform a json_decode on input attributes? Or might that happen for other parameters within vue as well?

File: plugins/ScheduledReports/vue/src/ListReports/ListReports.vue [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+              :id="`downloadReportForm_${report.idreport}`"
+              :action="linkTo({
+                module: 'API',
+                segment: null,
+                method: 'ScheduledReports.generateReport',
+                idReport: report.idreport,
+                outputType: downloadOutputType,
+                language: language,
+                format: ['html', 'csv', 'tsv'].indexOf(report.format) !== -1
+                  ? report.format : undefined,
```
Comment (sgiehl) [inline plugins/ScheduledReports/vue/src/ListReports/ListReports.vue]: We maybe could use `original` here as default. Currently downloading a SMS report won't work, as the API request doesn't contain a format and uses `XML` as default.

File: plugins/ScheduledReports/vue/src/AddReport/AddReport.vue [hunk 6]
```diff
@@ -0,0 +1,546 @@
...
+          :model-value="report.idsegment"
+          @update:model-value="$emit('change', { prop: 'idsegment', value: $event })"
+          :options="savedSegmentsById"
+        >
+          <template v-slot:inline-help>
+            <div
+              id="reportSegmentInlineHelp"
+              class="inline-help-node"
+              v-if="segmentEditorActivated"
+              v-html="reportSegmentInlineHelp"
```
Comment (sgiehl) [inline plugins/ScheduledReports/vue/src/AddReport/AddReport.vue]: this misses a `$sanitize`

File: plugins/ScheduledReports/vue/src/AddReport/AddReport.vue [hunk 7]
```diff
@@ -0,0 +1,546 @@
...
+          </div>
+        </div>
+      </div>
+      <SaveButton
+        :value="saveButtonTitle"
+        @confirm="$emit('submit')"
+      />
+      <div
+        class="entityCancel"
+        v-html="entityCancelText"
```
Comment (sgiehl) [inline plugins/ScheduledReports/vue/src/AddReport/AddReport.vue]: this misses a `$sanitize`

**Review by sgiehl — CHANGES_REQUESTED** [review]
Left a couple of comments for smaller improvements / fixes. Otherwise looks good.

File: plugins/ScheduledReports/templates/index.twig (hunk 4 above)
Comment (diosmosis) [reply plugins/ScheduledReports/templates/index.twig]: There's a JSON.parse() in the code that handles `vue-entry` in `piwikHelper`. Vue doesn't exist in HTML, all props in Vue are already instantiated, unserialized JavaScript values.

**Comment by sgiehl** [comment]
@diosmosis just found one last regression on this one. If the description of the report contains a special char like `<` or `"` it will be displayed encoded. Guess the "problem" is that the title is already stored encoded in the database. Before it was using `rawSafeDecoded` in twig. Not sure if we already have something similar in vue to use.

**Comment by diosmosis** [comment]
@sgiehl :+1: we'd just decode it client side. Vue just sets textContent/innerText so its safe (unless using v-html which uses innerHTML).

**Review by sgiehl — APPROVED** [review]

**Pull request merged**